x-pack/plugin/text-structure/src/main/java/org/elasticsearch/xpack/textstructure/transport/TransportFindMessageStructureAction.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.textstructure.transport;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.ActionFilters;
+import org.elasticsearch.action.support.HandledTransportAction;
+import org.elasticsearch.common.inject.Inject;
+import org.elasticsearch.common.util.concurrent.EsExecutors;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.transport.TransportService;
+import org.elasticsearch.xpack.core.textstructure.action.FindMessageStructureAction;
+import org.elasticsearch.xpack.core.textstructure.action.FindStructureResponse;
+import org.elasticsearch.xpack.textstructure.structurefinder.TextStructureFinder;
+import org.elasticsearch.xpack.textstructure.structurefinder.TextStructureFinderManager;
+import org.elasticsearch.xpack.textstructure.structurefinder.TextStructureOverrides;
+import org.elasticsearch.xpack.textstructure.structurefinder.TimeoutChecker;
+
+import java.util.ArrayList;
+
+import static org.elasticsearch.threadpool.ThreadPool.Names.GENERIC;
+
+public class TransportFindMessageStructureAction extends HandledTransportAction<FindMessageStructureAction.Request, FindStructureResponse> {
+
+    private final ThreadPool threadPool;
+
+    @Inject
+    public TransportFindMessageStructureAction(TransportService transportService, ActionFilters actionFilters, ThreadPool threadPool) {
+        super(
+            FindMessageStructureAction.NAME,
+            transportService,
+            actionFilters,
+            FindMessageStructureAction.Request::new,
+            EsExecutors.DIRECT_EXECUTOR_SERVICE
```

Comment:
Since `doExecute` executes its entire contents in the generic threadpool, there is a more elegant way to do this:

```suggestion
            threadPool.executor(GENERIC)
```

I can see you copied this code from `TransportFindStructureAction`. I'm not sure why it was done this way in that file as well. Maybe due to how the file evolved over time - it probably started off doing the analysis on the network thread and then got changed to fork to a different thread but without thinking of the best practice way to do that.

You could make the same change to the existing `TransportFindStructureAction`.

Reply:
Thanks, fixed.

Just to make sure: this means I can completely get rid of the 
```
threadPool.executor(GENERIC).execute(() -> { ... });
```
right, and just execute the code synchronously (because it's already on the generic threadpool)

Reply:
Another question: should I apply the same pattern to `TransportTestGrokPatternAction` (including letting it extend `HandledTransportAction` instead of `TransportAction`)?